File: services/cart-service/src/main/java/com/flashsale/cart/service/CartService.java
package com.flashsale.cart.service;

import com.flashsale.cart.api.dto.CartItemResponse;
import com.flashsale.cart.api.dto.CartResponse;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

@Service
public class CartService {

    private static final Duration CART_TTL = Duration.ofDays(7);
    private final StringRedisTemplate redis;

    public CartService(StringRedisTemplate redis) {
        this.redis = redis;
    }

    private String key(String userId) {
        return "cart:" + userId;
    }

    public CartResponse getCart(String userId) {
        Map<Object, Object> entries = redis.opsForHash().entries(key(userId));

        List<CartItemResponse> items = entries.entrySet().stream()
                .map(e -> new CartItemResponse(String.valueOf(e.getKey()), Integer.parseInt(String.valueOf(e.getValue()))))
                .sorted(Comparator.comparing(CartItemResponse::sku))
                .toList();

        return new CartResponse(userId, items);
    }

    public CartResponse addItem(String userId, String sku, int qty) {
        String k = key(userId);
        // atomic: increment qty if exists else create
        redis.opsForHash().increment(k, sku, qty);
        redis.expire(k, CART_TTL);
        return getCart(userId);
    }

    public CartResponse setQty(String userId, String sku, int qty) {
        if (qty <= 0) throw new IllegalArgumentException("qty must be >= 1");
        String k = key(userId);
        redis.opsForHash().put(k, sku, String.valueOf(qty));
        redis.expire(k, CART_TTL);
        return getCart(userId);
    }

    public CartResponse removeItem(String userId, String sku) {
        redis.opsForHash().delete(key(userId), sku);
        return getCart(userId);
    }

    public void clearCart(String userId) {
        redis.delete(key(userId));
    }
}

File: ui/flashsale-ui/src/app/core/cart.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { CartResponse } from './models';

@Injectable({ providedIn: 'root' })
export class CartService {
  constructor(private api: ApiService) {}

  getCart(userId: string) {
    return this.api.get<CartResponse>(`/cart/${userId}`);
  }

  addItem(userId: string, sku: string, qty: number) {
    return this.api.post<CartResponse>(`/cart/${userId}/items`, { sku, qty });
  }

  setQty(userId: string, sku: string, qty: number) {
    return this.api.put<CartResponse>(`/cart/${userId}/items/${sku}?qty=${qty}`);
  }

  removeItem(userId: string, sku: string) {
    return this.api.delete<CartResponse>(`/cart/${userId}/items/${sku}`);
  }

  clearCart(userId: string) {
    return this.api.delete<void>(`/cart/${userId}`);
  }
}